## Exploring Ebay Car Sales Data
In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. We've made a few modifications from the original dataset that was uploaded to Kaggle:

We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)
The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.
- The aim of this project is to clean the data and analyze the included used car listings. You'll also - become familiar with some of the unique benefits jupyter notebook provides for pandas.

Let's start by importing the libraries we need and reading the dataset into pandas.

### 1. Cleaning Columns Names
1. Import the pandas and NumPy libraries
2. Read the autos.csv CSV file into pandas, and assign it to the variable name autos.
    - Try without specifying any encoding (which will default to UTF-8)
    - If you get an encoding error, try the next two most popular encodings (Latin-1 and Windows-1252)       until you are able to read the file without error.
3. Create a new cell with just the variable autos and run this cell.
    - A neat feature of jupyter notebook is its ability to render the first few and last few values of       any pandas object.
4. Use the DataFrame.info() and DataFrame.head() methods to print information about the autos dataframe, as well as the first few rows.
    - Write a markdown cell briefly describing your observations.

In [1]:
import pandas as pd

autos = pd.read_csv("autos.csv", encoding="Latin-1")
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

From the work we did in the last screen, we can make the following observations:

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just     replace spaces with underscores.

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

1. Use the DataFrame.columns attribute to print an array of the existing column names.
2. Copy that array and make the following edits to columns names:
    - yearOfRegistration to registration_year
    - monthOfRegistration to registration_month
    - notRepairedDamage to unrepaired_damage
    - dateCreated to ad_created
    - The rest of the columnn names from camelcase to snakecase.
3. Assign the modified column names back to the DataFrame.columns attribute.
4. Use DataFrame.head() to look at the current state of the autos dataframe.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
column_name = {
    "yearOfRegistration": "registration_year",
    "monthOfRegistration": "registration_month",
    "notRepairedDamage": "unrepaired_damage",
    "dateCreated": "ad_created"
}

autos.rename(columns=column_name, inplace=True)
autos.rename(str.lower, axis="columns", inplace=True)

In [6]:
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### 2. Initial Exploration and Cleaning
Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for: - Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis. - Examples of numeric data stored as text which can be cleaned and converted.

The following methods are helpful for exploring the data: - `DataFrame.describe()` (with `include='all'` to get both categorical and numeric columns) - `Series.value_counts()` and `Series.head()` if any columns need a closer look.

- Use DataFrame.describe() to look at descriptive statistics for all columns.
- You likely found that the price and odometer columns are numeric values stored as text. For each column:
    - Remove any non-numeric characters.
    - Convert the column to a numeric dtype.
    - Use DataFrame.rename() to rename the column to odometer_km.

In [7]:
autos.describe()

registration_year       powerps  registration_month  nrofpictures  \
count       50000.000000  50000.000000        50000.000000       50000.0   
mean         2005.073280    116.355920            5.723360           0.0   
std           105.712813    209.216627            3.711984           0.0   
min          1000.000000      0.000000            0.000000           0.0   
25%          1999.000000     70.000000            3.000000           0.0   
50%          2003.000000    105.000000            6.000000           0.0   
75%          2008.000000    150.000000            9.000000           0.0   
max          9999.000000  17700.000000           12.000000           0.0   

         postalcode  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [8]:
autos["price"] = autos["price"].str.replace("[$,]", "").astype(float)
autos["odometer"] = autos["odometer"].str.replace("[mk,]", "").astype(float)
autos.rename(columns={"odometer": "odometer_km"}, inplace=True)

### 3. Exploring the Odometer and Price Columns
From the last screen, we learned that there are a number of text columns where almost all of the values are the same (`seller` and `offer_type`). We also converted the `price` and `odometer` columns to numeric types and renamed `odometer` to `odometer_km`.

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the `odometer_km` and `price` columns. Here's the steps we'll take:

- Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.
- We'll use:
    - `Series.unique().shape` to see how many unique values
    - `Series.describe()` to view min/max/median/mean etc
    - `Series.value_counts()`, with some variations:
        - chained to `.head()` if there are lots of values.
        - Because `Series.value_counts()` returns a series, we can use `Series.sort_index()` with `ascending=` `True` or `False` to view the highest and lowest values with their counts (can also chain to `head()` here).
    - When removing outliers, we can do `df[(df["col"] > x ) & (df["col"] < y )]`, but it's more readable to use `df[df["col"].between(x,y)]`

- For each of the odometer_km and price columns:
    - Use the techniques above to explore the data
    - If you find there are outliers, remove them and write a markdown paragraph explaining your decision.
    - After you have removed the outliers, make some observations about the remaining values.

In [9]:
autos["price"].unique().shape

(2357,)

In [10]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [11]:
autos["price"].value_counts().sort_index()

0.0           1421
1.0            156
2.0              3
3.0              1
5.0              2
8.0              1
9.0              1
10.0             7
11.0             2
12.0             3
13.0             2
14.0             1
15.0             2
17.0             3
18.0             1
20.0             4
25.0             5
29.0             1
30.0             7
35.0             1
40.0             6
45.0             4
47.0             1
49.0             4
50.0            49
55.0             2
59.0             1
60.0             9
65.0             5
66.0             1
              ... 
151990.0         1
155000.0         1
163500.0         1
163991.0         1
169000.0         1
169999.0         1
175000.0         1
180000.0         1
190000.0         1
194000.0         1
197000.0         1
198000.0         1
220000.0         1
250000.0         1
259000.0         1
265000.0         1
295000.0         1
299000.0         1
345000.0         1
350000.0         1
999990.0         1
999999.0    

In [12]:
autos = autos[autos["price"].between(100, 151990)]

In [13]:
autos["odometer_km"].unique().shape

(13,)

In [14]:
autos["odometer_km"].describe()

count     48205.000000
mean     125957.058396
std       39498.090902
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [15]:
autos["odometer_km"].value_counts().sort_index(ascending=True)

5000.0        753
10000.0       243
20000.0       753
30000.0       776
40000.0       813
50000.0      1008
60000.0      1152
70000.0      1214
80000.0      1412
90000.0      1733
100000.0     2100
125000.0     5037
150000.0    31211
Name: odometer_km, dtype: int64

### 4. Exploring the date columns
Let's now move on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:
```
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website
```
Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [16]:
autos[['datecrawled','ad_created','lastseen']][0:5]

datecrawled           ad_created             lastseen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

You'll notice that the first 10 characters represent the day (e.g. 2016-03-12). To understand the date range, we can extract just the date values, use Series.value_counts() to generate a distribution, and then sort by the index.

To select the first 10 characters in each column, we can use Series.str[:10]:

In [17]:
print(autos['datecrawled'].str[:10].head())

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: datecrawled, dtype: object


1. Use the workflow we just described to calculate the distribution of values in the date_crawled, ad_created, and last_seen columns (all string columns) as percentages.
    - To include missing values in the distribution and to use percentages instead of counts, chain the Series.value_counts(normalize=True, dropna=False) method.
    - To rank by date in ascending order (earliest to latest), chain the Series.sort_index() method.

In [18]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033109
2016-03-10    0.032009
2016-03-11    0.032922
2016-03-12    0.036760
2016-03-13    0.017052
2016-03-14    0.035308
2016-03-15    0.034063
2016-03-16    0.029955
2016-03-17    0.031159
2016-03-18    0.013567
2016-03-19    0.033627
2016-03-20    0.037859
2016-03-21 

In [19]:
autos['lastseen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.001079
2016-03-06    0.004315
2016-03-07    0.005435
2016-03-08    0.007323
2016-03-09    0.009563
2016-03-10    0.010642
2016-03-11    0.012405
2016-03-12    0.023794
2016-03-13    0.008879
2016-03-14    0.012634
2016-03-15    0.015870
2016-03-16    0.016430
2016-03-17    0.028109
2016-03-18    0.007302
2016-03-19    0.015766
2016-03-20    0.020641
2016-03-21    0.020537
2016-03-22    0.021367
2016-03-23    0.018587
2016-03-24    0.019770
2016-03-25    0.019106
2016-03-26    0.016658
2016-03-27    0.015538
2016-03-28    0.020828
2016-03-29    0.022301
2016-03-30    0.024707
2016-03-31    0.023836
2016-04-01    0.022861
2016-04-02    0.024873
2016-04-03    0.025122
2016-04-04    0.024541
2016-04-05    0.125008
2016-04-06    0.222010
2016-04-07    0.132165
Name: lastseen, dtype: float64

### 5. Dealing with Incorrect Registration Year Data
One thing that stands out from the exploration we did in the last screen is that the `registration_year` column contains some odd values:

- The minimum value is `1000`, before cars were invented
- The maximum value is `9999`, many years into the future
Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

1. Decide which the highest and lowest acceptable values are for the registration_year column.
2. Remove the values outside those upper and lower bounds and calculate the distribution of the remaining values using Series.value_counts(normalize=True).

In [20]:
autos["registration_year"].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000041
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000041
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000041
1943    0.000021
1948    0.000021
1950    0.000021
1951    0.000041
1952    0.000021
1953    0.000021
1954    0.000041
1955    0.000041
1956    0.000083
1957    0.000041
1958    0.000083
1959    0.000124
1960    0.000456
1961    0.000124
1962    0.000083
1963    0.000166
1964    0.000249
          ...   
2000    0.064371
2001    0.054517
2002    0.051385
2003    0.055866
2004    0.055990
2005    0.060388
2006    0.055347
2007    0.047153
2008    0.045825
2009    0.043149
2010    0.032922
2011    0.033565
2012    0.027072
2013    0.016575
2014    0.013712
2015    0.007800
2016    0.024852
2017    0.028690
2018    0.009709
2019    0.000041
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000062
5911    0.000021
6200    0.000021
8888    0.0000

In [21]:
autos = autos[autos["registration_year"].between(1910, 2019)]
autos["registration_year"].value_counts(normalize=True)

2000    0.064396
2005    0.060412
1999    0.059748
2004    0.056012
2003    0.055888
2006    0.055369
2001    0.054539
2002    0.051405
1998    0.048541
2007    0.047171
2008    0.045843
2009    0.043166
1997    0.039949
2011    0.033578
2010    0.032935
2017    0.028701
1996    0.028120
2012    0.027083
2016    0.024862
1995    0.024758
2013    0.016582
2014    0.013718
1994    0.012991
2018    0.009712
1993    0.008716
2015    0.007803
1992    0.007616
1991    0.007014
1990    0.006890
1989    0.003549
          ...   
1977    0.000436
1969    0.000394
1975    0.000374
1965    0.000353
1964    0.000249
1963    0.000166
1961    0.000125
1959    0.000125
1937    0.000083
1962    0.000083
1958    0.000083
1956    0.000083
1955    0.000042
1951    0.000042
2019    0.000042
1934    0.000042
1957    0.000042
1910    0.000042
1941    0.000042
1954    0.000042
1948    0.000021
1953    0.000021
1950    0.000021
1927    0.000021
1943    0.000021
1929    0.000021
1939    0.000021
1938    0.0000

### 6. Exploring Price by Brand
One of the analysis techniques we learned in this course is aggregation. When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the brand column.

If you recall in an earlier mission, we explored how to use loops to perform aggregation. Here's what the process looks like:
```
- Identify the unique values we want to aggregate by
- Create an empty dictionary to store our aggregate data
- Loop over the unique values, and for each:
    - Subset the dataframe by the unique values
    - Calculate the mean of whichever column we're interested in
    - Assign the val/mean to the dict as k/v.
```

1. Explore the unique values in the brand column, and decide on which brands you want to aggregate by.
    - You might want to select the top 20, or you might want to select those that have over a certain percentage of the total values (e.g. > 5%).
    - Remember that Series.value_counts() produces a series with index labels, so you can use Series.index attribute to access the labels, should you wish.
2. Write a short paragraph describing the brand data, and explaining which brands you've chosen to aggregate on.
3. Create an empty dictionary to hold your aggregate data.
    - Loop over your selected brands, and assign the mean price to the dictionary, with the brand name as the key.
    - Print your dictionary of aggregate data, and write a paragraph analyzing the results.

In [22]:
temp = autos[["brand", "price"]]
brand_lst = list(autos["brand"].value_counts().index)[:20]
brand_mean = {}
top_six_brand = ["bmw", "mercedes_benz", "opel", "audi", "volkswagen", "ford"]
for name in brand_lst:
    if name in top_six_brand:
        result = temp[temp["brand"] == name].mean()[0]
        brand_mean[name] = result
brand_mean

{'audi': 9219.534249879402,
 'bmw': 8181.499236641222,
 'ford': 3756.9919547079858,
 'mercedes_benz': 8498.783953287197,
 'opel': 2971.6082375478927,
 'volkswagen': 5366.253433330086}

### 7. Sorting Aggregate Data in a DataFrame
In the last screen, we aggregated across brands to understand mean price. We observed that in the top 6 brands, there's a distinct price gap.

- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

- it's difficult to compare more than two aggregate series objects if we want to extend to more columns
- we can't compare more than a few rows from each series object
- we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Instead, we can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly. To do this, we'll need to learn two pandas methods:

- pandas series constructor
- pandas dataframe constructor

Here's an example of the series constructor that uses the `brand_mean_prices` dictionary:

In [23]:
bmp_series = pd.Series(brand_mean)
print(bmp_series)

audi             9219.534250
bmw              8181.499237
ford             3756.991955
mercedes_benz    8498.783953
opel             2971.608238
volkswagen       5366.253433
dtype: float64


The keys in the dictionary became the index in the series object. We can then create a single-column dataframe from this series object. We need to use the `columns` parameter when calling the dataframe constructor (which accepts a array-like object) to specify the column name (or the column name will be set to `0` by default):

In [24]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
audi           9219.534250
bmw            8181.499237
ford           3756.991955
mercedes_benz  8498.783953
opel           2971.608238
volkswagen     5366.253433

1. Use the loop method from the last screen to calculate the mean mileage and mean price for each of the top brands, storing the results in a dictionary.
2. Convert both dictionaries to series objects, using the series constructor.
3. Create a dataframe from the first series object using the dataframe constructor.
4. Assign the other series as a new column in this dataframe.

In [26]:
temp = autos[["brand", "odometer_km"]]

mileage_mean = {}
for name in top_six_brand:
    result = temp[temp["brand"] == name].mean()[0]
    mileage_mean[name] = result
mileage_mean

{'audi': 129634.58755426918,
 'bmw': 132884.54198473282,
 'ford': 124360.8462455304,
 'mercedes_benz': 131116.9982698962,
 'opel': 129512.45210727969,
 'volkswagen': 129056.19947404305}

In [27]:
mileage_series = pd.Series(mileage_mean)
df_mileage = pd.DataFrame(mileage_series, columns=["mean_mileage"])
df_mileage

mean_mileage
audi           129634.587554
bmw            132884.541985
ford           124360.846246
mercedes_benz  131116.998270
opel           129512.452107
volkswagen     129056.199474